Classes/Private/Manager/Audio/OCTAudioEngine.m

```diff
@@ -83,459 +38,162 @@ - (instancetype)init
-                                            kAudioUnitScope_Global,
-                                            kInputBus,
-                                            &callbackStruct,
-                                            sizeof(callbackStruct));
-
-    if (status != noErr) {
-        [self fillError:error
-               withCode:status
-            description:@"Registering Input Callback"
-          failureReason:@"Failed to register input callback"];
+#if TARGET_OS_IPHONE
+    AVAudioSession *session = [AVAudioSession sharedInstance];
+
+    if (! ([session setCategory:AVAudioSessionCategoryPlayAndRecord error:error] &&
+           [session setPreferredSampleRate:kDefaultSampleRate error:error] &&
+           [session setMode:AVAudioSessionModeVoiceChat error:error] &&
+           [session setActive:YES error:error])) {
         return NO;
     }
-    return YES;
-}
+#endif
 
-- (BOOL)registerOutputCallBack:(NSError **)error;
-{
-    AURenderCallbackStruct callbackStruct;
-    callbackStruct.inputProc = outputRenderCallBack;
-    callbackStruct.inputProcRefCon = (__bridge void *)(self);
-    OSStatus status = _AudioUnitSetProperty(self.ioUnit,
-                                            kAudioUnitProperty_SetRenderCallback,
-                                            kAudioUnitScope_Global,
-                                            kOutputBus,
-                                            &callbackStruct,
-                                            sizeof(callbackStruct));
-
-    if (status != noErr) {
-        [self fillError:error
-               withCode:status
-            description:@"Registering output Callback"
-          failureReason:@"Failed to register output callback"];
+    [self makeQueues:error];
+
+    if (! (self.outputQueue && self.inputQueue)) {
         return NO;
     }
-    return YES;
-}
 
+    OCTAudioEngine *__weak welf = self;
+    self.inputQueue.sendDataBlock = ^(void *data, OCTToxAVSampleCount samples, OCTToxAVSampleRate rate, OCTToxAVChannels channelCount) {
+        OCTAudioEngine *aoi = welf;
 
-OSStatus inputRenderCallBack(void *inRefCon,
-                             AudioUnitRenderActionFlags  *ioActionFlags,
-                             const AudioTimeStamp    *inTimeStamp,
-                             UInt32 inBusNumber,
-                             UInt32 inNumberFrames,
-                             AudioBufferList *ioData)
-{
-    OCTAudioEngine *engine = (__bridge OCTAudioEngine *)(inRefCon);
+        if (aoi.enableMicrophone) {
+            [aoi.toxav sendAudioFrame:data
+                          sampleCount:samples
+                             channels:channelCount
+                           sampleRate:rate
+                             toFriend:aoi.friendNumber
+                                error:nil];
+        }
+    };
+    [self.outputQueue updateSampleRate:(Float64)self.outputSampleRate numberOfChannels:(UInt32)self.outputNumberOfChannels error:nil];
```

Comment:
outputQueue can take OCTToxAVSampleRate and OCTToxAVChannels as parameters and do a cast itself. That way we'll have OCT types in more places